bin/aws_write_creds.sh

```diff
@@ -25,16 +21,21 @@ if [ ${#missing_auth_vars[@]} -ne 0 ]
 then
     echo "Did not find values for:" 
     printf ' %q\n' "${missing_vars[@]}"
-    echo "Will assume they are in credentials file"
+    echo "Will assume they are in credentials file or not needed"
 else
     echo "Creating credentials file"
     # Create the directory....
     mkdir -p ~/.aws
     CREDS=~/.aws/credentials
     echo "[default]"                                    >  $CREDS
     echo "aws_access_key_id=$AWS_ACCESS_KEY_ID"         >> $CREDS 
-    echo "aws_secret_access_key=$AWS_SECRET_ACCESS_KEY" >> $CREDS 
-    echo "aws_session_token=$AWS_SESSION_TOKEN"         >> $CREDS 
+    echo "aws_secret_access_key=$AWS_SECRET_ACCESS_KEY" >> $CREDS
+    # This is if we have non-temp credentials...
+    if [[ -z "${AWS_SESSION_TOKEN+x}" ]]; then
```

Comment:
Checking my understanding here, we just want to test if the env var has a value.  If the env var has a value, we don't want to actually interpolate it, so we replace it with `x` as described [here](https://stackoverflow.com/questions/46891981/what-does-argumentx-mean-in-bash) and the reason we do so is because the var may contain sensitive information.

Reply:
Partially; the other half of this is that we want to check to see if the variable is set or not.  

In this example, if AWS_SESSION_TOKEN is set, that expression evaluates to x, which is then tested with -z (which returns true on a zero-length string), and found false so (in this case) we know we have a value in that variable.

However, if the variable is unset then that expression evaluates to nothing, which when tested by -z is true (since it is a zero-length string).

Why do we go through this trouble? One thing is that if we have `set -o unbound` in the script our script will blow up if it hits an unbound variable (which is the point, but it can bite one in the ass at times). The other thing is that doing a -z on just the variable will allow for two conditions - the unset variable we care about and a variable set to an empty string (which we may not want).

I think this makes a good argument for rewriting this in python... ;)